retriever/lib/engine_tools.py

```diff
@@ -101,37 +104,102 @@ def reset_retriever(scope="all", ask_permission=True):
             print("can't find script {scp}".format(scp=scope))
 
 
-def json2csv(input_file, output_file=None, header_values=None, encoding=ENCODING):
-    """Convert Json file to CSV.
-
-    Function is used for only testing and can handle the file of the size.
-    """
+def json2csv(input_file,
+             output_file=None,
+             header_values=None,
+             encoding=ENCODING,
+             row_key=None):
+    """Convert Json file to CSV."""
     file_out = open_fr(input_file, encoding=encoding)
     # set output file name and write header
     if output_file is None:
         output_file = os.path.splitext(os.path.basename(input_file))[0] + ".csv"
     csv_out = open_fw(output_file, encoding=encoding)
     if os.name == 'nt':
-        outfile = csv.DictWriter(csv_out,
-                                 dialect='excel',
-                                 escapechar="\\",
-                                 lineterminator='\n',
-                                 fieldnames=header_values)
+        outfile = csv.writer(csv_out,
+                             dialect='excel',
+                             escapechar="\\",
+                             lineterminator='\n')
     else:
-        outfile = csv.DictWriter(csv_out,
-                                 dialect='excel',
-                                 escapechar="\\",
-                                 fieldnames=header_values)
-    raw_data = json.loads(file_out.read())
-    outfile.writeheader()
+        outfile = csv.writer(csv_out, dialect='excel', escapechar="\\")
 
-    for item in raw_data:
-        outfile.writerow(item)
+    raw_data = json.loads(file_out.read(), object_pairs_hook=OrderedDict)
+
+    raw_data, header_values = walker(raw_data,
+                                     row_key=row_key,
+                                     header_values=header_values,
+                                     rows=[],
+                                     normalize=False)
+
+    if isinstance(raw_data[0], dict):
+        raw_data = [list(row.values()) for row in raw_data]
+    else:
+        raw_data = [row.tolist() for row in raw_data]
+    if header_values:
+        outfile.writerow(header_values)
+    outfile.writerows(raw_data)
     file_out.close()
     subprocess.call(['rm', '-r', input_file])
     return output_file
 
 
+def walker(dictionary, row_key=None, header_values=None, rows=[], normalize=False):
+    """
+    Extract rows of data from json datasets
+    """
+    #  Handles the simple case, where row_key and column_key are not required
+    if not (row_key or header_values):
+        if isinstance(dictionary, dict):
+            rows = pd.DataFrame([dictionary]).values
+            header_values = dictionary.keys()
+            return rows, header_values
+        elif isinstance(dictionary, list):
```

Comment:
if `dictionary` can also be of type list maybe we should use some other name in place of `dictionary` here as it can be confusing.

Reply:
Done